### PR TITLE
Add github actions to build, test, and deploy site automatically

### PR DESCRIPTION
**Pull request by pradeepnschrodinger**
## Description

I'm adding two GitHub action workflows to automate
1. building and testing changes made to the repo
2. deploying our FDT site to gh-pages whenever a release is published (see: http://schrodinger.github.io/fixed-data-table-2/)

For more information on github actions, see [docs](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#overview.).

### Deployment
I often forget to publish our static site to gh-pages after doing a release.
I wanted a way to automate the process so I don't have to do this (and remember to do this) every time a release happens.

### CI
We had travis integrated to our repo for CI, but it stopped working for some reason **since May 25th last year** (based on our [commit history](https://github.com/schrodinger/fixed-data-table-2/commits/master)).
I only realized this some days back. 
Luckily we were always manually running `yarn run test` so our mainlines are still in a good state.
I could look into setting up / fixing travis once more, but I figured that I could do just use GitHub actions to automate testing instead.

## How Has This Been Tested?

I tested the workflows in [my fork of FDT](https://github.com/pradeepnschrodinger/fixed-data-table-2/tree/pradeepnschrodinger_v1.2.x).
See the various actions being dispatched there: https://github.com/pradeepnschrodinger/fixed-data-table-2/actions

### Starting logs of build-test workflow
Screen Shot 2022-06-13 at 11 53 52 PM

### Ending logs of build-test workflow
Screen Shot 2022-06-13 at 11 55 01 PM

### Graph of deploy workflow
Screen Shot 2022-06-13 at 11 53 13 PM

### CODE REVIEW

**Comment by pradeepnschrodinger** [comment]
Now that this PR is created, checkout https://github.com/schrodinger/fixed-data-table-2/actions to see the new workflows in action.

File: .github/workflows/build-and-run-tests.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# This workflow will be run whenever a new commit is pushed or when a PR is opened up.
+# Guide: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions
+
+name: Build and Run Tests
+
+on:
+  # run this for every push (CI)
+  push:
+  # also run this for PR
+  pull_request:
```
Comment (pradeepnschrodinger) [inline .github/workflows/build-and-run-tests.yml:11]: Using both the `push` and `pull_request` made this workflow run twice when I opened up this PR...
Not sure what's the best approach here...
The intent was to make sure our own commits, as well PRs be tested.

File: .github/workflows/deploy-gh-pages.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and site, and run the deploy script.
+# The deploy script will publish the build to gh-pages.
+# This workflow will be run whenever a new release is published.
+
+# NOTE (pradeep): I'm having this workflow bail out if the release doesn't target v1.2.x. 
+# This is because we don't have a way (yet) to support multiple FDT versions/docs  be deployed at the same time,
+# and hence we choose to deploy only the most stable release.
```
Comment (pradeepnschrodinger) [inline .github/workflows/deploy-gh-pages.yml:7]: In the near/distant future, I'm thinking of having something like [lodash's docs site](https://lodash.com/docs/4.17.15) which runs multiple major releases, and has a extremely simple dropdown at the top to switch between versions.

File: .github/workflows/deploy-gh-pages.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      run: |
+        echo "RELEASE_TAG is: " $RELEASE_TAG
+        if [[ $RELEASE_TAG =~ ^v1\.2\.[0-9]+$ ]]; 
+        then
+          echo "This is a stable release!"
+          echo "::set-output name=is_stable::true"
+        else
+          echo "This is not a stable release!"
+          echo "::set-output name=is_stable::false"
+        fi
```
Comment (pradeepnschrodinger) [inline .github/workflows/deploy-gh-pages.yml:45]: In the future, we'll be customizing the deploy steps to support multiple releases.

File: .github/workflows/deploy-gh-pages.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    - name: Install dependencies
+      run: yarn install --frozen-lockfile # do a clean install of all of our dependencies
+
+    # build our static contents
+    - name: Build static site!
+      run: yarn run site-build
+
+    # publish our site to gh-pages
+    - name: Deploy to gh-pages
+      uses: peaceiris/actions-gh-pages@v3
```
Comment (pradeepnschrodinger) [inline .github/workflows/deploy-gh-pages.yml:78]: I wasn't able to figure out how to deploy to gh-pages manually.

I tried using the [gh-pages](https://www.npmjs.com/package/gh-pages) package but it looks it requires me to authenticate, and I'm way too unfamiliar with security stuff.
So I'm instead using https://github.com/peaceiris/actions-gh-pages, which seems well used and also worked well here.

File: package.json [hunk 5]
```diff
@@ -93,7 +93,7 @@
   "bugs": {
     "url": "https://github.com/schrodinger/fixed-data-table-2/issues"
   },
-  "homepage": "http://schrodinger.github.io/fixed-data-table-2",
+  "homepage": "https://schrodinger.github.io/fixed-data-table-2",
```
Comment (pradeepnschrodinger) [inline package.json:96]: I'm changing this to `https`.
Once this PR is ready to merged in, I'll also follow the steps in https://docs.github.com/en/pages/getting-started-with-github-pages/securing-your-github-pages-site-with-https#enforcing-https-for-your-github-pages-site

File: .github/workflows/deploy-gh-pages.yml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }} # authenticate the action to our workflow
```
Comment (wcjordan) [inline .github/workflows/deploy-gh-pages.yml:80]: where does this get configured?

**Comment by wcjordan** [comment]
Nice, this looks very cool.  Thanks for setting this up.

**Review by wcjordan — APPROVED** [review]

File: .github/workflows/deploy-gh-pages.yml (hunk 6 above)
Comment (pradeepnschrodinger) [reply .github/workflows/deploy-gh-pages.yml:80]: This is a special secret that's automatically created and configured for each workflow run.

From https://docs.github.com/en/actions/security-guides/automatic-token-authentication#about-the-github_token-secret:
> At the start of each workflow run, GitHub automatically creates a unique GITHUB_TOKEN secret to use in your workflow. You can use the GITHUB_TOKEN to authenticate in a workflow run.
> 
> When you enable GitHub Actions, GitHub installs a GitHub App on your repository. The GITHUB_TOKEN secret is a GitHub App installation access token. You can use the installation access token to authenticate on behalf of the GitHub App installed on your repository. The token's permissions are limited to the repository that contains your workflow.

Comment (wcjordan) [reply .github/workflows/deploy-gh-pages.yml:80]: Very cool.  Thanks for the explanation & documentation
